local/local_test.go

```diff
@@ -84,91 +84,236 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
+						h.AssertEq(t, inspect.Os, daemonOS)
+						h.AssertEq(t, inspect.Architecture, "amd64")
+						h.AssertEq(t, inspect.RootFS.Type, "layers")
+
+						h.AssertEq(t, img.Found(), true)
+					})
 				})
-			})
 
-			when("base image and daemon os/architecture match", func() {
-				it("uses the base image architecture/OS", func() {
-					img, err := local.NewImage(newTestImageName(), dockerClient, local.FromBaseImage(runnableBaseImageName))
-					h.AssertNil(t, err)
-					h.AssertNil(t, img.Save())
-					defer h.DockerRmi(dockerClient, img.Name())
+				when("base image and daemon architecture do not match", func() {
+					it("uses the base image architecture/OS", func() {
+						armBaseImageName := "busybox@sha256:50edf1d080946c6a76989d1c3b0e753b62f7d9b5f5e66e88bef23ebbd1e9709c"
+						expectedArmArch := "arm64"
+						expectedOSVersion := ""
+						if daemonOS == "windows" {
+							// windows/arm nanoserver image
+							armBaseImageName = "mcr.microsoft.com/windows/nanoserver@sha256:29e2270953589a12de7a77a7e77d39e3b3e9cdfd243c922b3b8a63e2d8a71026"
+							expectedArmArch = "arm"
+							expectedOSVersion = "10.0.17763.1040"
+						}
 
-					imgOS, err := img.OS()
-					h.AssertNil(t, err)
-					h.AssertEq(t, imgOS, daemonOS)
+						h.PullIfMissing(t, dockerClient, armBaseImageName)
 
-					inspect, _, err := dockerClient.ImageInspectWithRaw(context.TODO(), img.Name())
-					h.AssertNil(t, err)
-					h.AssertEq(t, inspect.Os, daemonOS)
-					h.AssertEq(t, inspect.Architecture, "amd64")
-					h.AssertEq(t, inspect.RootFS.Type, "layers")
+						img, err := local.NewImage(
+							newTestImageName(),
+							dockerClient,
+							local.FromBaseImage(armBaseImageName),
+						)
+						h.AssertNil(t, err)
+						h.AssertNil(t, img.Save())
+						defer h.DockerRmi(dockerClient, img.Name())
 
-					h.AssertEq(t, img.Found(), true)
+						imgArch, err := img.Architecture()
+						h.AssertNil(t, err)
+						h.AssertEq(t, imgArch, expectedArmArch)
+
+						imgOSVersion, err := img.OSVersion()
+						h.AssertNil(t, err)
+						h.AssertEq(t, imgOSVersion, expectedOSVersion)
+					})
 				})
 			})
 
-			when("base image and daemon architecture do not match", func() {
-				it("uses the base image architecture", func() {
-					armBaseImageName := "arm64v8/busybox@sha256:50edf1d080946c6a76989d1c3b0e753b62f7d9b5f5e66e88bef23ebbd1e9709c"
-					expectedArmArch := "arm64"
-					expectedOSVersion := ""
-					if daemonOS == "windows" {
-						// this nanoserver windows/arm image exists and pulls. Not sure whether it works for anything else.
-						armBaseImageName = "mcr.microsoft.com/windows/nanoserver@sha256:29e2270953589a12de7a77a7e77d39e3b3e9cdfd243c922b3b8a63e2d8a71026"
-						expectedArmArch = "arm"
-						expectedOSVersion = "10.0.17763.1040"
-					}
+			when("#WithPlatform", func() {
+				when("base image and platform architecture/OS do not match", func() {
+					it("uses the base image architecture/OS, ignoring platform", func() {
```

Comment:
Should we just fail in this case? Given that we can't sensibly reconcile the two options?

Reply:
I could be swayed but it felt like it starts to deviate from `remote`, which in this impl just wraps `ggcr`'s [behavior](https://github.com/google/go-containerregistry/blob/51f01e739161bf35d663240a8ca404777d7d9378/pkg/v1/remote/descriptor.go#L152-L156), which is to only consider platform for manifest list refs (which are never a thing for `local` since it's refs are unique).

My general take on `WithPlatform()` had been primarily to assign values to a new image when nothing is inherited - when either no `FromBaseImage()` is given or when the `FromBaseImage()` image is not found. It's secondary purpose is for remote's special edge-case when FromBaseImage() needs a manifest list tie-breaker. I wasn't yet treating it as a constraint or validator.

Reply:
I think this gets tricky because we have `WithPlatform` doing double-duty for `remote` (1. as a new image value setter, 2. as a tie-breaker for manifest lists). If we don't want `imgutil` consumers to need to know ahead of time wether a remote ref will be a manifest list or image manifest, and swap in our out `WithPlatform` accordingly, I think we need to keep passing-through image manifests so consumers can decide what to do to with them. And because we're stuck with that for `remote`, I think we need to stay consistent in `local`.

I could be swayed though, happy to hear counter arguments.